Crash Site Arrakis

Jayefuu opened the pull request:
- Drier than desert. No trees or oases so wood and fish must be bought from the market.
- Placing roboports attracts sandworms. They will spawn every 6 to 12 minutes to attack each roboport so networks must be kept small or protected well.
- Repair cars. Upon placing repair packs in a car it can fire a repair beam at nearby damaged entities.

Jayefuu committed:
Crash Site Arrakis

Jayefuu committed:
Fixed lua check fails

grilledham commented on map_gen/maps/crash_site/features/repair_cars.lua:
```diff
@@ -0,0 +1,95 @@
+    car_entities = tbl.car_entities
+end)
+
+local function register_car(entity)
+    if not entity then
+        return
+    end
+    if not entity.valid then
+        return
+    end
+    if entity.name ~= 'car' then
+        return
+    end
+    car_entities[entity.unit_number] = entity
+    script.register_on_entity_destroyed(entity)
+end
+
+Event.add(defines.events.on_entity_destroyed, function(event)
+    car_entities[event.unit_number] = nil
+end)
+
+Event.add(defines.events.on_robot_built_entity, function(event)
+    register_car(event.created_entity)
+end)
+
+Event.add(defines.events.on_built_entity, function(event)
+    register_car(event.created_entity)
+end)
+
+Event.add(defines.events.on_entity_cloned, function(event)
+    register_car(event.destination)
+end)
```
I wonder if the <https://lua-api.factorio.com/latest/events.html#on_player_driving_changed_state> event would be better to use (Same for rocket tanks).

Assuming the event does what I think it does, it would mean a smaller list of entities to iterate through and we could remove the get_driver check.

Jayefuu replied:
That makes good sense. It will be more overhead when players/passengers get in and out of vehicles but less to check per tick.

grilledham commented on map_gen/maps/crash_site/features/repair_cars.lua:
```diff
@@ -0,0 +1,95 @@
+                -- if the damaged entity is not the car. Cars can heal other cars but not themselves.
+                -- if the damaged entity is not a character
+                -- if the entity is not moving. Vehicles must have a speed of 0, entities that can't move will be nil
+                if (entity ~= car and (entity.speed == 0 or entity.speed == nil) and entity.name ~= 'character' and not turrets[entity.name]) or (turrets[entity.name] and entity.active == true) then
```
Normal grumble about repeating the same table lookups 😛

Jayefuu replied:
I think I've fixed this in the upcoming commit. I created vars before the if statement then used those in the if statement conditions.

grilledham commented on map_gen/maps/crash_site/features/repair_cars.lua:
```diff
@@ -0,0 +1,95 @@
+                    end
+                    entity.health = entity.health + repair_amount
+                    stack.drain_durability(repair_amount)
+                    goto continue -- break out because we only want to heal one entity
```
Could probably use `break` here but this is fine.

Jayefuu replied:
Thanks.

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,114 @@
+                entity.position.x + math.random(max_worm_spawn_radius * -1, max_worm_spawn_radius),
+                entity.position.y + math.random(max_worm_spawn_radius * -1, max_worm_spawn_radius)
+            }
+            worm_position = s.find_non_colliding_position(index, worm_position, 5, 1)
```
`find_non_colliding_position` can return `nil`, so we should check.

Jayefuu replied:
Fixed

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,114 @@
+            -- For the appropriate worm for each evolution region, spawn some accompanying biters to attack the roboport
+            for worm, _ in pairs(sandworm_biters) do
+                if worm == index then
+                    for biters, data in pairs(sandworm_biters[worm]) do
```
Isn't `sandworm_biters[worm]` the value you discarded in the above for loop?

Jayefuu replied:
Ahhh yes. I've added biters instead of _ and sandworm_biters[worm]. That helped me understand tables better, thanks.

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,114 @@
+            -- For the appropriate worm for each evolution region, spawn some accompanying biters to attack the roboport
+            for worm, _ in pairs(sandworm_biters) do
+                if worm == index then
+                    for biters, data in pairs(sandworm_biters[worm]) do
```
The key called `biters` is really the name of a single biter, so I think it should be just `biter`.

Jayefuu replied:
Fixed.

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,114 @@
+                                amount = math.floor(amount)
+                            end
+                        end
+                        for _ = amount,1,-1 do
```
If you don't use the index, isn't this the same as `for _ = 1, amount do`

Jayefuu replied:
Yes. Thanks. Fixed. Been programming in other languages a lot and couldn't remember the correct syntax. Used the first thing I found in google.

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,114 @@
+                            end
+                        end
+                        for _ = amount,1,-1 do
+                           local pos = s.find_non_colliding_position(biters, worm_position, 5, 1)
```
`find_non_colliding_position` could return nil

Jayefuu replied:
Updated.

grilledham commented on map_gen/maps/crash_site/presets/arrakis.lua:
```diff
@@ -0,0 +1,39 @@
+ScenarioInfo.set_map_name('Crashsite Arrakis')
+ScenarioInfo.set_map_description('Capture outposts and defend against the biters. Even drier than desert, sandworms roam the desert and will attack roboports on sight.')
+ScenarioInfo.add_map_extra_info(
+    '- Arrakis is even drier than crash site Desert.\n- Sandworms are attracted to the vibration caused by roboports and will spawn intermittently to neutralise this threat to their peace\n- Cars have repair beams\n- Outposts have enemy turrets defending them.\n- Outposts have loot and provide a steady stream of resources.\n- Outpost markets to purchase items and outpost upgrades.\n- Capturing ou ...
```
Using multi line string `[[...]]` see <https://www.lua.org/pil/2.4.html> might make these easier to read.

Jayefuu replied:
All updated in next commit.

grilledham reviewed (commented):
I had a quick look, haven't play tested yet.

Jayefuu committed:
Updated as per requests

grilledham commented on map_gen/maps/crash_site/features/repair_cars.lua:
```diff
@@ -0,0 +1,89 @@
+    local entity = event.entity
+    if (entity.name ~= 'car') or not entity or not entity.valid then
+        return
+    else
```
Don't need an `else` after a `return`

grilledham commented on map_gen/maps/crash_site/presets/arrakis.lua:
```diff
@@ -0,0 +1,46 @@
+    .. '- Outposts have loot and provide a steady stream of resources.\n'
+    .. '- Outpost markets to purchase items and outpost upgrades.\n'
+    .. '- Capturing outposts increases evolution.\n- Reduced damage by all player weapons, turrets, and ammo.\n'
+    .. '- Biters have more health and deal more damage.\n- Biters and spitters spawn on death of entities.'
```
That's not what I meant by multi line strings 😄 
```suggestion
[[
- Arrakis is even drier than crash site Desert.
- Sandworms are attracted to the vibration caused by roboports and will spawn intermittently to neutralise this threat to their peace
- Cars have repair beams
- Outposts have enemy turrets defending them.
- Outposts have loot and provide a steady stream of resources.
- Outpost markets to purchase items and outpost upgrades.
- Capturing outposts increases evolution.\n- Reduced damage by all player weapons, turrets, and ammo.
- Biters have more health and deal more damage.\n- Biters and spitters spawn on death of entities.]]
```

Jayefuu commented:
So sandworms aren't working properly in live. I need to re-test with some debug.

Jayefuu commented:
Ok they're working fine, but the non_colliding radius needs turning up for in base

Jayefuu committed:
Changed balance and updated as per requests

sandworms.lua
- Added feedback to player that places the roboport
- Added floating text *thump* as warning when roboport placed and 30 seconds before attack
- Changed worm/biter spawn balance
- Fixed failed lua check
- Changed spawn time to 5 to 15 minutes instead of 6 to 20 minutes

repair_cars.lua:
- Removed surplus else statement as requested
- Added ability for tanks to repair. Cars were very underwhelming past medium worms, their health is too low to be useful.

others:
- Fixed multiline strings in ScenarioInfo

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,133 @@
+
+Event.add(defines.events.on_entity_cloned, function(event)
+    if event.created_entity.name == 'roboport' then
+        start_worm_attacks(event.created_entity)
```
on_entity_cloned event doesn't have a created_entity property.
Also when is on_entity_cloned raised, is it a map editor thing?

grilledham replied:
I think maybe you missed this one, It doesn't seem to be resolved.

Jayefuu replied:
It seems this is only raised when cloning entities from one surface to another. As such I've removed it.

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,133 @@
+
+local thump_text_callback
+thump_text_callback = Token.register(function(entity)
+    local s = entity.surface
```
Need an entity.valid check.

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,133 @@
+end)
+
+local function start_worm_attacks(entity)
+    if not entity then
```
Is this supposed to be an entity.valid check instead?

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,133 @@
+
+local worm_callback
+worm_callback = Token.register(function(entity)
+    if entity then -- stops the callback if the roboport has been removed
```
Is this supposed to be an entity.valid check instead?

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,133 @@
+end
+
+Event.add(defines.events.on_robot_built_entity, function(event)
+    if event.created_entity.name == 'roboport' then
+        start_worm_attacks(event.created_entity)
+    end
+end)
+
+Event.add(defines.events.on_built_entity, function(event)
+    if event.created_entity.name == 'roboport' then
```
The build events need to check the entity is valid. 
There can be multiple event handlers per event and a handler that runs before this handler might have destroyed the entity which would make it invalid. That was a difficult bug to track down we once had ~3years ago.

Jayefuu committed:
Added entity validity checks

Jayefuu commented:
I think I've resolved everything you raised. :)

grilledham commented on map_gen/maps/crash_site/features/vehicle_repair_beams.lua:
```diff
@@ -0,0 +1,89 @@
+
+Event.add(defines.events.on_player_driving_changed_state, function(event)
+    local entity = event.entity
+    if (entity.type ~= 'car') or not entity or not entity.valid then
```
Probably should do the nil check before trying to access the property.
```suggestion
    if not entity or not entity.valid or (entity.type ~= 'car') then
```
Edit: Should probably do the valid check before the type check too.

grilledham commented on map_gen/maps/crash_site/features/vehicle_repair_beams.lua:
```diff
@@ -0,0 +1,89 @@
+                -- if the damaged entity is not the car. Cars can heal other cars but not themselves.
+                -- if the damaged entity is not a character
+                -- if the entity is not moving. Vehicles must have a speed of 0, entities that can't move will be nil
+                if (entity ~= car and (entity.speed == 0 or entity.speed == nil) and entity.name ~= 'character' and not turrets[entity.name]) or (turrets[entity.name] and entity.active == true) then
```
We did discuss this on discord. Are there any entities that are inactive other than the turrets?
I'm wondering if the turret check is needed at all. I think you might only need the active check.

Jayefuu replied:
Belts, splitters, underground belts, chests and other entities can have active==false, and I want to repair those, so I need to check that it's a turret and that it's active so that it repairs active turrets but not while they're on turret cooldown.

grilledham replied:
Ok, that's good to know, I think this is fine then.

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,136 @@
+end
+
+Event.add(defines.events.on_robot_built_entity, function(event)
+    if event.created_entity.name == 'roboport' and event.created_entity.valid then
```
Needs an entity.valid check for the same reason as the on_built_entity event does.

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,136 @@
+end)
+
+Event.add(defines.events.on_built_entity, function(event)
+    if event.created_entity.name == 'roboport' and event.created_entity.valid then
```
The valid check should be before accessing the properties of the entity.

Jayefuu committed:
Updated as per requests

Jayefuu commented:
Ok I think I actually got them all this time @grilledham  :D

Jayefuu committed:
Updated crash site upgrade toasts to include player name

It's handy to know who's doing the upgrading so I've added their name to all the crash site toasts. This branch depends on the arrakis/dunes one as in that branch I moved some things into a features folder.

Jayefuu committed:
Added discord prints to rocket tank and airstrike spawn market upgrades

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,130 @@
+                s.create_entity {name = index, position = worm_position, force = "enemy"}
+            end
+            -- For the appropriate worm for each evolution region, spawn some accompanying biters to attack the roboport
+            for worm, biters in pairs(sandworm_biters) do
```
Why do you need to iterate this table, can't you just get the biters table using `local biters = sandworm_biters[index]`?

grilledham commented on map_gen/maps/crash_site/features/sandworms.lua:
```diff
@@ -0,0 +1,130 @@
+Event.add(defines.events.on_built_entity, function(event)
+    if event.created_entity.valid and event.created_entity.name == 'roboport' then
+        start_worm_attacks(event.created_entity)
+        local player = game.get_player(event.player_index)
```
Probably not needed, but a player not nil or invalid check would be good.

Jayefuu committed:
Merge pull request #1147 from Refactorio/improved_outpost_toast_messages

Updated crash site upgrade toasts to include player name
